### ISSUE TITLE
useQuery returns responses from cache

### ISSUE THREAD

**Issue by Kabilan35** [marcin-piela/react-fetching-library#99]
useQuery hook keeps returning response from stored cache even when I set cache:"no-cache" in the Action
I use RN states to refresh I my app, it makes a call and returns the cachedResponse and not fresh data. It works fine when I kill and launch my app but does not help while refreshing

**Comment by rlaffers** [comment]
Yes, I also encountered this behavior. Here is my analysis:

To be more specific, useQuery() keeps returning response stored in its local state on re-renders. This local state is not synchronized with the cached state stored in cache provider (and local state is created even if you do not use cache provider at all - which is your case).
This is due to the use of `useReducer` in [useQuery.ts](https://github.com/marcin-piela/react-fetching-library/blob/master/src/hooks/useQuery/useQuery.ts#L17):
```
const [state, dispatch] = useReducer(responseReducer as ResponseReducer<T>, {
    loading: cachedResponse ? false : initFetch,
    response: cachedResponse ? cachedResponse : { error: false },
  });```

The upshot of the above is this:
1. When you do not use cache, you are still getting the response fetched once, and then stored in local state. All subsequent re-renders will get this response. This response will be stored until your component is unmounted.
2. When you use cache with a timeout (the response should become stale after a certain amount of time), the timing out is not respected. You'll keep getting the locally stored response even if the cache-stored response has timed out and was removed from the cache.

The first behavior is kind of OK, since you probably do not want to re-fetch on every re-render. When you want to re-fetch, the workaround is to use the `query` function returned by the useQuery hook.
The second behavior is undesirable. Local state should not trump the cache. There should probably be an effect inside useQuery which resets the local state when it detects that the cached response is no longer valid. Something like
```
useEffect(() => {
  if (!cachedResponse) {  // pseudo-code. This check needs to be more sophisticated
    dispatch({ type: RESET });
  }  
}, [cachedResponse, dispatch])
```
Or something like that.

**Comment by jerrysu** [comment]
The use of `useReducer` in `useQuery` also causes the hook to override/reuse the cached response for other actions during re-renders. This causes the incorrect data to be displayed.

For example:

```javascript
function Foo({ query }) {
  const { payload } = useQuery(SearchActionCreator(query));
  if (payload) {
    return (
      <pre>
        {query}
        {JSON.stringify(payload)}
      </pre>
    );
  }
  return null;
}
```

After `payload` is loaded in the code above and the `query` prop changes, `payload` is the cached response of the previous value of `query`. This is definitely a bug.

`useSuspenseQuery` doesn't have this issue since it doesn't use `useReducer` internally.

A workaround if `useQuery` must be used is to force the element to re-render by keying it with the value used by the action creator (e.g., `<Foo key={query} query={query} />`).

**Comment by jerrysu** [comment]
@marcin-piela: A fix for the bug in `useQuery` would be to reset the reducer whenever the `action` changes.

Simply adding `handleReset()` to  https://github.com/marcin-piela/react-fetching-library/blob/master/src/hooks/useQuery/useQuery.ts#L22-L33:

```
  useEffect(() => {
    isMounted.current = true;

    handleReset();

    if (initFetch && !cachedResponse) {
      handleQuery();
    }

    return () => {
      isMounted.current = false;
      handleAbort();
    };
  }, [convertActionToBase64(action)]);
```

**Comment by marcin-piela** [comment]
Fixed in 1.7.5 version
